cpp/Embeddings.cc

```diff
@@ -0,0 +1,56 @@
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+
+#include "Embeddings.hh"
+#include "finalfusion.h"
+
+class Embeddings::EmbeddingsImpl {
+  friend Embeddings;
+
+public:
+  EmbeddingsImpl(std::string const &filename, bool mmap) {
+    ff_embeddings embeddings;
+    if (mmap) {
+      embeddings = ff_mmap_embeddings(filename.c_str());
+    } else {
+      embeddings = ff_read_embeddings(filename.c_str());
+    }
+
+    if (embeddings == nullptr) {
+      throw std::runtime_error(std::string(ff_error()));
+    }
+    d_inner = embeddings;
+  }
+
+  ~EmbeddingsImpl() { ff_free_embeddings(d_inner); }
+
+private:
+  size_t dimensions() { return ff_embeddings_dims(d_inner); }
+
+  std::vector<float> embedding(std::string const &word) {
+    float *raw_embedding = ff_embedding_lookup(d_inner, word.c_str());
+    if (raw_embedding == nullptr) {
+      return std::vector<float>();
+    }
+
+    size_t dims = dimensions();
+    std::vector<float> embedding(raw_embedding, raw_embedding + dims);
+    free(raw_embedding);
+    return embedding;
+  }
+
+  ff_embeddings d_inner;
```

Comment:
Forgot another thing in my previous review: classes that have bare pointer members should be non-copyable or provide a copy-constructor and assignment implementation. Otherwise, assignment/copy construction will result in a double-free.

See http://www.icce.rug.nl/documents/cplusplus/cplusplus09.html#l175 for more information.

Reply:
First of all, I hope that I'm not misunderstanding this, but iirc, wrapping `ff_embeddings` in a `unique_ptr` would also solve the issue, right?

I wrapped `ff_embeddings` in a `unique_ptr` but can't quite figure out how to do so in a clean manner. The constructor of `unique_ptr` expects a pointer `*ff_embeddings` if I define the pointer as `std::unique_ptr<ff_embeddings>`. 

Our C-API returns `ff_embeddings` from the `read` and `mmap` methods, so I didn't figure out a way to construct `unique_ptr` without casting `ff_embeddings` to `ff_embeddings*`, which should be save since `ff_embeddings` is actually a pointer anyways:

```C++
std::unique_ptr<ff_embeddings>((ff_embeddings *) ff_read_embeddings(path))
```

Since all our ffi methods expect `ff_embeddings` rather than `ff_embeddings*`, I also need to cast the pointer to get access to the methods, e.g.:
```C++
ff_embedding_lookup((ff_embeddings) d_inner.get(), word.c_str())
```

Is there a simpler way to do this? If that's not a valid solution to the problem, then I'll make the class non-copyable.

Reply:
> First of all, I hope that I'm not misunderstanding this, but iirc, wrapping `ff_embeddings` in a `unique_ptr` would also solve the issue, right?

Indeed, `unique_ptr` is not copy constructable, nor assignable.

> I wrapped `ff_embeddings` in a `unique_ptr` but can't quite figure out how to do so in a clean manner. The constructor of `unique_ptr` expects a pointer `*ff_embeddings` if I define the pointer as `std::unique_ptr<ff_embeddings>`.

`ff_embeddings` is already a pointer type, so you'd want std::unique_ptr<ff_embeddings_t>.

> Our C-API returns `ff_embeddings` from the `read` and `mmap` methods, so I didn't figure out a way to construct `unique_ptr` without casting `ff_embeddings` to `ff_embeddings*`, which should be save since `ff_embeddings` is actually a pointer anyways:

That's not correct, `ff_embeddings*` is a pointer to a pointer to `ff_embeddings_t` (two levels of indirection, rather than one).

Reply:
Thanks for clearing that up, I was trying to use `ff_embeddings_t` earlier but got thrown off by other errors in the code. I pushed a version with `std::unique_ptr<ff_embeddings_t>`